### PR TITLE
feat: create dep tree from project.assets.json

### PR DESCRIPTION
**Pull request by lwywoo**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/dotnet-deps-parser/blob/master/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
Builds a dependency tree from project.assets.json

#### Where should the reviewer start?
lib/parsers/project-assets-json-parser.ts

#### What are the relevant tickets?
JIRA: https://snyksec.atlassian.net/browse/BST-646

### CODE REVIEW

File: lib/parsers/project-assets-json-parser.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  const topLevelDepTree: PkgTree = {
+    dependencies: {},
+    name,
+    version,
+  };
+  return topLevelDepTree;
+}
+
+export function getDependencyTreeFromProjectAssetsJson(manifestFile: ProjectAssetsJsonManifest,
+                                                       targetFrameWork): PkgTree {
```
Comment (miiila) [inline lib/parsers/project-assets-json-parser.ts]: This long space looks bit weird :-)

File: lib/parsers/project-assets-json-parser.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  };
+  project: {
+    restore: {
+      projectName: string,
+    },
+    version: string,
+  };
+}
+
+export function getProjectNameForProjectAssetsJson(manifestFile) {
```
Comment (miiila) [inline lib/parsers/project-assets-json-parser.ts]: Why do we need to export these functions? Is it just because of the testing?

Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: Yup this is just for the tests

File: lib/parsers/project-assets-json-parser.ts (hunk 1 above)
Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: The linter enforced as it was over max-line length :(

File: lib/parsers/project-assets-json-parser.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+function buildRootPackageTree(name, version) {
+  const depTree: PkgTree = {
+    dependencies: {},
+    name,
+    version,
+  };
+  return depTree;
+}
+
+function buildPackageTreeForDep(name, version) {
```
[the remark this answers is not in the document]
Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: Removed

File: lib/parsers/project-assets-json-parser.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+export function getDependencyTreeFromProjectAssetsJson(manifestFile: ProjectAssetsJsonManifest,
+                                                       targetFrameWork): PkgTree {
+  const projectName = getProjectNameForProjectAssetsJson(manifestFile);
+  const projectVersion = getProjectVersionForProjectAssetsJson(manifestFile);
+  const depTree = buildRootPackageTree(projectName, projectVersion);
+
+  const topLevelDeps = Object.keys(_.get(manifestFile, ['targets', targetFrameWork], {}));
+  for (const topLevelDep of topLevelDeps) {
+    const topLevelDepName = topLevelDep.split('/')[0];
```
[the remark this answers is not in the document]
Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: Thanks, updated

File: lib/index.ts [hunk 5]
```diff
@@ -38,6 +44,16 @@ function buildDepTreeFromProjectJson(manifestFileContents: string, includeDev =
   return getDependencyTreeFromProjectJson(manifestFile, includeDev);
 }
 
+// TODO: Figure out what to do about devDeps
+function buildDepTreeFromProjectAssetsJson(manifestFileContents: string, targetFramework?: string): PkgTree {
+  if (!targetFramework) {
+    throw new Error('Missing required parameters for buildDepTreeFromProjectAssetsJson()');
```
Comment (orsagie) [inline lib/index.ts]: Since it's just one parameter, can you say `missing targetFramework for ...`

File: lib/parsers/index.ts [hunk 6]
```diff
@@ -367,11 +368,9 @@ export function getTargetFrameworksFromProjectConfig(manifestFile) {
...
 
 export function getTargetFrameworksFromProjectJson(manifestFile) {
-  const targetFrameworksResult: string[] = [];
   return Object.keys(_.get(manifestFile, 'frameworks', {}));
 }
 
 export function getTargetFrameworksFromProjectAssetsJson(manifestFile) {
-  const targetFrameworksResult: string[] = [];
-  return Object.keys(_.get(manifestFile, 'project.frameworks', {}));
+  return Object.keys(_.get(manifestFile, 'targets', {}));
```
Comment (orsagie) [inline lib/parsers/index.ts]: In snyk-nuget-plugin we get the framework from project.frameworks (https://github.com/snyk/snyk-nuget-plugin/blob/master/lib/nuget-parser/dotnet-core-parser.js#L88)

Why did you change it here?

File: lib/parsers/project-assets-json-parser.ts (hunk 1 above)
Comment (orsagie) [reply lib/parsers/project-assets-json-parser.ts]: You can change it to be
```suggestion
export function getDependencyTreeFromProjectAssetsJson(
    manifestFile: ProjectAssetsJsonManifest,
    targetFrameWork,
): PkgTree {
```

File: lib/index.ts (hunk 5 above)
Comment (lwywoo) [reply lib/index.ts]: Update error message :)

File: lib/parsers/project-assets-json-parser.ts (hunk 1 above)
Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: 👍 updated

File: lib/parsers/index.ts (hunk 6 above)
Comment (lwywoo) [reply lib/parsers/index.ts]: In order to get the deps from the project.assets.json, I needed the targets which have the format ".NETCoreApp,Version=v2.2" instead of "netcoreapp2.2" which is the format from project.frameworks. Is it better to be retrieve the framework from project.frameworks and then do some transfomation to get the targetFramework format I require?
`{
    "version": 3,
    "targets": {
      ".NETCoreApp,Version=v2.2": {
        "log4net/2.0.8": {
          "type": "package",
          "dependencies": {
            "System.AppContext": "4.1.0",`

Comment (orsagie) [reply lib/parsers/index.ts]: In CLI the required format is `netcoreapp2.2` so that is why it uses project.frameworks, if for SCM we use the other format, then better to not have to create our own targetframework name interpreter.

File: lib/parsers/project-assets-json-parser.ts (hunk 2 above)
Comment (miiila) [reply lib/parsers/project-assets-json-parser.ts]: In my opinion, than we don't need to test it at all. If it's a part of public interface just because of the testing, it usually means something strange with the design.

In this case, I believe it's enough to test that whole tree deep equals to the expectation. WDYT?

Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: We can remove it if we don't think that the testing the function is valuable, it was useful for me to write the test for the function with the stubs in place first and then extending to test for the entire expected tree.

Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: Removed the exporting `getProjectNameForProjectAssetsJson` and `getProjectVersionForProjectAssetsJson` and removed the unit tests

File: lib/parsers/project-assets-json-parser.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  project: {
+    restore: {
+      projectName: string,
+    },
+    version: string,
+  };
+}
+
+export function getProjectNameForProjectAssetsJson(manifestFile) {
+  return _.get(manifestFile, ['project', 'restore', 'projectName'], {});
```
[the remark this answers is not in the document]
Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: Im not sure how to use a variable with _get and dot notation .e.g like for `topLevelDeps = Object.keys(_.get(manifestFile, ['targets', targetFrameWork], {}));` where targetFrameworks is passed in. 
If you can provide an example, I am happy to change it.

[the remark this answers is not in the document]
Comment (miiila) [reply lib/parsers/project-assets-json-parser.ts]: In case you mentioned, the only way would be:
```
topLevelDeps = Object.keys(_.get(manifestFile, `targets.${targetFrameWork}`, {}));
```

But I don't think it's more readable :-) My original comment was for the case without variable.

[the remark this answers is not in the document]
Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: Cool, I can update the variables without variables with dot notation if that is more readable :)

[the remark this answers is not in the document]
Comment (lwywoo) [reply lib/parsers/project-assets-json-parser.ts]: Tried updating to use dotnotation but was not able to retrieve the project name and version :(
